src/Shelly.hs

```diff
@@ -325,22 +325,26 @@ runCommand handles st exe args = findExe exe >>= \fullExe ->
 #if defined(mingw32_HOST_OS)
       fp
 #else
-      _fp
+      fp
 #endif
-      = do
-        mExe <- whichEith exe
-        case mExe of
-          Right execFp -> return execFp
-          -- windows looks in extra places besides the PATH, so just give
-          -- up even if the behavior is not properly specified anymore
-          --
-          -- non-Windows < 7.8 has a bug for read-only file systems
-          -- https://github.com/yesodweb/Shelly.hs/issues/56
-          -- it would be better to specifically detect that bug
+      = case fp of
+          -- If the 'FilePath' contains a more articulated path than just
+          -- an executable, don't try to find it.
```

Comment:
Should this rather check whether there is a directory component in `exe`? What about `bla/hello.sh` instead of `./hello.sh`?  Currently you would accept e.g. `.bla/hello.sh` or `.hello.sh`.
I think checking for _starting with dot_ isn't the correct criterion.

Reply:
Yes, I take your point this is too simplistic. I will have to think about this a little bit.

Reply:
@andreasabel Aha! After a more thorough look, I have finally nailed down exactly why the regression happened. We have this code, which is responsible for locating the file, in the guts of `whichEith`:

```hs
    whichFull fp = do
      (trace . mappend "which " . toTextIgnore) fp >> whichUntraced
      where
        whichUntraced | isAbsolute fp          = checkFile
                      | dotSlash splitOnDirs   = checkFile
                      | length splitOnDirs > 0 = lookupPath  >>= leftPathError
                      | otherwise              = lookupCache >>= leftPathError

        splitOnDirs = splitDirectories fp
        dotSlash ("./":_) = True
        dotSlash _ = False
```

As you can see, we are already accounting for `dotSlash`, so why things fail? Well, it turns out that both `system-filepath`  and `filepath` have a `splitDirectory` function, but hey behave differently:

```
> import qualified Filesystem.Path.CurrentOS as SFP
> import qualified Data.Text as T
> import System.FilePath as FP
> FP.splitDirectories "./test/data/hello.sh"
[".","test","data","hello.sh"]
> SFP.splitDirectories (SFP.fromText $ T.pack "./test/data/hello.sh")
[FilePath "./",FilePath "test/",FilePath "data/",FilePath "hello.sh"]
> 
```

As you can see the former doesn't include the "./", so the pattern match on `dotSlash` fails, and here is our regression.

Reply:
Excellent analysis! Thanks, @adinapoli!